mecha.zig

```diff
@@ -29,25 +30,111 @@ pub fn Parser(comptime _T: type) type {
         pub const mapConst = mecha.mapConst;
         pub const map = mecha.map;
         pub const opt = mecha.opt;
+
+        const Self = @This();
+
+        pub fn match(
+            self: Self,
+            alloc: mem.Allocator,
+            str: []const u8,
+            result: Result(T).Pass,
+        ) !void {
+            const res = try self.parse(alloc, str);
+            switch (res) {
+                .ok => try expectResult(T, .{ .ok = result }, res),
+                .err => res.reportError(),
+            }
+        }
+
+        pub fn fail(
+            self: Self,
+            alloc: mem.Allocator,
+            str: []const u8,
+            pos: usize,
+        ) !void {
+            switch (try self.parse(alloc, str)) {
+                .ok => return error.TestExpectedError,
+                .err => |e| try testing.expectEqual(pos, e.pos),
+            }
+        }
     };
 }
 
-/// The result of a successful parse
+/// The span of the input that was covered by a parser.
+pub const Span = struct {
+    start: usize,
+    len: usize,
+};
+
+pub inline fn span(start: usize, len: usize) Span {
+    return .{ .start = start, .len = len };
+}
+
+/// The result of a parse where `ok` corresponds to a successful parse
+/// and `err` denotes a failure. The result will be placed in `value`
+/// and `rest` will contain the unparsed input. Result will be annotated
+/// with  a `span` containing the starting position of the result in the
+/// input string, as well as its length. Finally, `pos` will contain
+/// the position where the parser stopped and the next parser can pick up.
 pub fn Result(comptime T: type) type {
-    return struct {
-        pub const Value = T;
+    return union(enum) {
+        ok: Pass,
+        err: Fail,
+
+        const Pass = struct {
+            pub const Value = T;
+
+            value: T,
+            rest: []const u8 = "",
+            span: Span = span(0, 0),
+        };
```

Comment:
Would it be possible for us just to return position and value?

```suggestion
        const Pass = struct {
            pub const Value = T;

            value: T,
            pos: usize,
        };
```

Pretty sure this has all the information need

Reply:
Knowing the span of a token should make for better error reporting during type-checking. For example, enabling one to highlight the whole “token”. Just tracking position won’t do it.

Reply:
I’ll add tests to check. The parent parser should update the span start for results returned by child parsers.

Reply:
Unfortunately, my PR is not gonna work for token span tracking.

No span tracking is possible unless parsers take a context and this context is then propagated to child parsers (like in PR #73).

Only error position tracking is possible in the current version of the code since errors bubble up but only values of successful parses are stored and the rest of the result data discarded.